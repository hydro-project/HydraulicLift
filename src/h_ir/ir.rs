use std::rc::Rc;

use syn::{Expr, Ident};

use crate::{derive_hnode, utils::{
    debug::DebugStr, functional::Semigroup, pattern::{ExprPat, ScopePat}, scope::Scope, tagged::Tagged
}};


// :: value
#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct HOutput {
    pub input: HReturn,
    pub other: Option<Box<HOutput>>,
}
derive_hnode!(HOutput: Ident);

/// :: value
#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct HReturn {
    pub value: HExpr,
}
derive_hnode!(HReturn: Ident);

/// :: (value, scope)
#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub enum HExpr {
    Raw(Tagged<HExprRaw, Scope>),
    Await(HExprAwait),
    // A merge point
    Union(HExprUnion),
    /// A branch point
    Shared(HExprShared),
}
derive_hnode!(HExpr: ExprPat);


#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct HExprAwait(pub Box<HExpr>);
derive_hnode!(HExprAwait: ExprPat);

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct HExprRaw {
    pub expr: DebugStr<Expr>,
    pub input: HScope,
    pub scope_def: Scope,
}
derive_hnode!(HExprRaw: ExprPat);

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct HExprUnion(pub Box<HExpr>, pub Box<HExpr>);
derive_hnode!(HExprUnion: ExprPat);

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct HExprShared(pub Rc<HExpr>);
derive_hnode!(HExprShared: ExprPat);

/// :: scope
#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub enum HScope {
    Bind(Tagged<HBind, Scope>),
    Filter(HFilter),
    Input(Tagged<HInput, Scope>),
}
derive_hnode!(HScope: ScopePat);

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct HBind {
    pub id: Ident,
    pub value: Box<HExpr>,
}
derive_hnode!(HBind: ScopePat);

/// Filters for cond == expectation
#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct HFilter {
    pub expectation: bool,
    pub cond: Box<HExpr>,
}
derive_hnode!(HFilter: ScopePat);

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct HInput;
derive_hnode!(HInput: ScopePat);

impl HOutput {
    /// Creates a new output from the return value
    pub fn new(input: HReturn) -> Self {
        Self { input, other: None }
    }

    /// Creates a new output returning value
    pub fn ret(value: HExpr) -> Self {
        Self::new(HReturn::new(value))
    }
}

impl Semigroup for HOutput {
    fn concat(self, Self { input, other }: Self) -> Self {
        let move_one = Self {
            input,
            other: Some(Box::new(self)),
        };
        match other {
            Some(box rest) => move_one.concat(rest),
            None => move_one,
        }
    }
}

impl HReturn {
    pub fn new(value: HExpr) -> Self {
        Self { value }
    }
}

impl Semigroup for HExpr {
    fn concat(self, other: Self) -> Self {
        Self::Union(HExprUnion::new(self, other))
    }
}

impl HExprRaw {
    pub fn new(expr: Expr, input: HScope, scope_def: Scope) -> Self {
        Self {
            expr: expr.into(),
            input,
            scope_def,
        }
    }
}

impl HExprAwait {
    pub fn new(inner: HExpr) -> Self {
        Self(Box::new(inner))
    }
}

impl HExprUnion {
    pub fn new(e1: HExpr, e2: HExpr) -> Self {
        Self(Box::new(e1), Box::new(e2))
    }
}

impl HExprShared {
    pub fn new(inner: Rc<HExpr>) -> Self {
        Self(inner)
    }
}

impl HBind {
    pub fn new(id: Ident, value: HExpr) -> Self {
        Self {
            id,
            value: Box::new(value),
        }
    }
}

impl HFilter {
    pub fn new(expectation: bool, cond: HExpr) -> Self {
        Self {
            expectation,
            cond: Box::new(cond),
        }
    }
}
