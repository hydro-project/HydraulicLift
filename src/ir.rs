// use hydroflow_plus::ir::HfPlusNode;
// use quote::{quote, ToTokens};
// use syn::{Expr, Ident};

// use crate::{io::Scope, utils::{ident}};

// /// Return of a value to the outer hf+ program.
// /// :: (value, scope) -> value
// pub struct HReturn;

// /// Evaluation of an expression into a value.
// /// :: scope -> (value, scope)
// pub struct HExprRaw {
//     pub expr: Expr,
//     pub scope: Scope,
//     pub next: HExprConsumer,
// }

// /// Consumer of an expression.
// /// :: (value, scope) -> _
// pub enum HExprConsumer {
//     Map(HExprMap),
//     Bind(HExprBind),
//     Branch(HExprBranch),
//     BlockEnd(HBlockEnd),
//     Return(HReturn),
// }

// /// Mapping a function over an expression.
// /// :: (value, scope) -> (value, scope)
// pub struct HExprMap {
//     pub map_expr: Expr,
//     pub scope: Scope,
//     pub next: Box<HExprConsumer>,
// }

// /// Bind an expression value to a name in scope.
// /// :: (value, scope) -> scope U value
// pub struct HExprBind {
//     pub definition: Ident,
//     /// Does not contain ident
//     pub scope: Scope,
//     pub next: Box<HExprRaw>,
// }

// /// Branches based on a boolean expression value.
// /// :: (true, scope)  -> scope
// /// :: (false, scope) -> scope
// pub struct HExprBranch {
//     pub next_true: Box<HExprRaw>,
//     pub next_false: Box<HExprRaw>,
// }

// /// The end of a block.
// /// :: (value, old_scope) -> (value, new_scope)
// pub struct HBlockEnd {
//     pub old_scope: Scope,
//     pub new_scope: Scope,
//     pub next: Box<HExprConsumer>,
// }

// /// Entrypoint, turns input into expression value.
// /// :: value -> (value, [])
// pub struct HEntryPoint {
//     pub next: HExprConsumer
// }



// /*

// let x = 5;
// let y = if x > 5 {
//     return x + 1;
// } else {
//     x + 2
// }
// x + y


// =>

// PROBLEM: 

// expr(5) -> bind(x) -> expr(x > 5) -> tee()

// true? -> expr(x+1) -> return
// false? -> expr(x+2) -> ret_expr

// union() -> bind(y) -> expr(x+y)
// */
