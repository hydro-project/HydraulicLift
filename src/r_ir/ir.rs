use std::fmt::Debug;

use syn::{Expr, Ident};

use crate::utils::{debug::DebugStr, tagged::TagOut};

/// R AST - Extended syn AST which directly represents the Rust code.
/// New AST constructs are relevant for HF+ translation.
/// TODO: lift all HF+ AST relevant objects to the top level. All syn encapsulated objects should be raw rust


// I/O are tagged metadata corresponding to the inputs or outputs of a node.

#[derive(Debug, Clone)]
pub enum RExpr<I = (), O = ()> {
    If(RExprIf<I, O>),
    Block(RExprBlock<I, O>),
    Await(RExprAwait<I, O>),
    Raw(TagOut<RExprRaw<I>, O>), //TODO: expand
}

#[derive(Debug, Clone)]
pub struct RExprIf<I = (), O = ()> {
    pub cond_expr: Box<RExpr<I, O>>,
    pub then_expr: Box<RExpr<I, O>>,
    pub else_expr: Box<RExpr<I, O>>,
}

#[derive(Debug, Clone)]
pub struct RExprBlock<I = (), O = ()> {
    pub stmt: RStmt<I, O>,
    pub expr: Box<RExpr<I, O>>,
}

#[derive(Debug, Clone)]
pub struct RExprAwait<I = (), O = ()>(pub Box<RExpr<I, O>>);

#[derive(Debug, Clone)]
pub struct RExprRaw<I = ()> {
    pub expr: DebugStr<Expr>,
    pub scope: I,
}

#[derive(Debug, Clone)]
pub enum RStmt<I = (), O = ()> {
    Let(TagOut<RStmtLet<I, O>, O>),
    Return(RStmtReturn<I, O>),
    While(RStmtWhile<I, O>)
    // TODO: add expressions here?
}

#[derive(Debug, Clone)]
pub struct RStmtLet<I = (), O = ()> {
    pub id: Ident,
    pub is_mut: bool,
    pub value: Box<RExpr<I, O>>,
}

#[derive(Debug, Clone)]
pub struct RStmtReturn<I = (), O = ()> {
    pub value: Box<RExpr<I, O>>,
}

#[derive(Debug, Clone)]
pub struct RStmtWhile<I = (), O = ()> {
    pub cond: Box<RExpr<I, O>>,
    pub body: Box<RExpr<I, O>>,
}

impl<I, O> RExprIf<I, O> {
    pub fn new(cond_expr: RExpr<I, O>, then_expr: RExpr<I, O>, else_expr: RExpr<I, O>) -> Self {
        Self {
            cond_expr: Box::new(cond_expr),
            then_expr: Box::new(then_expr),
            else_expr: Box::new(else_expr),
        }
    }
}

impl<I, O> RExprBlock<I, O> {
    pub fn new(stmt: RStmt<I, O>, expr: RExpr<I, O>) -> Self {
        Self {
            stmt,
            expr: Box::new(expr),
        }
    }
}

impl<I, O> RExprAwait<I, O> {
    pub fn new(expr: RExpr<I, O>) -> Self {
        Self(Box::new(expr))
    }
}

impl<I> RExprRaw<I> {
    pub fn new(expr: Expr, input_scope: I) -> Self {
        Self {
            expr: expr.into(),
            scope: input_scope,
        }
    }
}

impl<I, O> RStmtLet<I, O> {
    pub fn new(id: Ident, is_mut: bool, value: RExpr<I, O>) -> Self {
        Self {
            id,
            is_mut,
            value: Box::new(value),
        }
    }
}

impl<I, O> RStmtReturn<I, O> {
    pub fn new(value: RExpr<I, O>) -> Self {
        Self {
            value: Box::new(value),
        }
    }
}

impl<I, O> RStmtWhile<I, O> {
    pub fn new(cond: RExpr<I, O>, body: RExpr<I, O>) -> Self {
        Self {
            cond: Box::new(cond),
            body: Box::new(body),
        }
    }
}

