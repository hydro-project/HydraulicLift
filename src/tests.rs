use hydroflow_plus::ir::HfPlusNode;

use crate::{compile, hfplus_ir::visualize::visualize};

macro_rules! test_compile {
    (let $input:ident = $hf_input:expr => $($body:tt)*) => {
        let hf = compile!(let $input = $hf_input => $($body)*);
        
        insta::assert_debug_snapshot!(hf);
        insta::assert_snapshot!(visualize(hf));
    };
}

#[test]
fn test_pipeline() {
    test_compile!(let hf_in = HfPlusNode::Placeholder => {
        let x = 1;
        let y = hf_in + 2;
        let z = x + y + 3;
        let o = z + 4;
        o + 5
    });
}

#[test]
fn test_return_simple() {
    test_compile!(let hf_in = HfPlusNode::Placeholder => {
        return hf_in;
    });
}

#[test]
fn test_return() {
    test_compile!(let hf_in = HfPlusNode::Placeholder => {
        let x = hf_in + 1;
        return x;
        x + 2 // this doesn't show up in the resulting HF+!
    });
}

#[test]
fn test_if_simple() {
    test_compile!(let hf_in = HfPlusNode::Placeholder => {
        if 1 == 2 {
            3
        } else {
            4
        }
    });
}

#[test]
fn test_async_simple() {
    test_compile!(let hf_in = HfPlusNode::Placeholder => {
        let x = async { hf_in };
        let y = x.await;
        y
    });
}

#[test]
fn test_globals_simple() {
    test_compile!(let hf_in = HfPlusNode::Placeholder => {
        // format and Some are global identifiers which should not be treated as locals
        let x = format!("hello");
        Some(x)
    });
}

#[test]
fn test_all() {
    test_compile!(let hf_in = HfPlusNode::Placeholder => {
        // Send a message asking if hf_in*2 is a registered id
        let id = hf_in * 2;
        let message = if id == 0 {
            "You asked for 0?".to_string()
        } else if check_exists(id).await {
            format!("Found id: {hf_in}")
        } else {
            return None;
            let abcd = panic!("This panic doesn't show up in the resulting hydroflow+!");
        };
        Some(message)
    });
}