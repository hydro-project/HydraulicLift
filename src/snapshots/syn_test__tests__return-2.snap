---
source: src/tests.rs
expression: visualize(hf)
---
node0=Placeholder;

node0.tee() 
	.map(| (hf_in) | { (hf_in) }) 
	.map(| (mut hf_in) | { let value = hf_in + 1 ; (value , ()) }) 
	.map(| (x , ()) | { (x) }) 
	.map(| (x) | { let value = x ; (value , ()) }) 
	.map(| (value , _) | { value })
